test/functional/mining.py

```diff
@@ -0,0 +1,336 @@
+            if txid not in block_txids:
+                ancestor_txids = node.getmempoolancestors(txid=txid, verbose=False)
+                for ancestor_txid in ancestor_txids:
+                    if ancestor_txid not in block_txids:
+                        txid_to_bump = txid
+                        break
+            if txid_to_bump is not None:
+                break
+        if txid_to_bump is None:
+            raise AssertionError("Not enough transactions to test ancestor feerate score (test bug?)")
+
+        # Calculate the ancestor feerate of the candidate transaction.
+        mempool_entry = node.getmempoolentry(txid_to_bump)
+        ancestor_size = mempool_entry['ancestorsize']
+        ancestor_fee = mempool_entry['ancestorfees']
+        self.log.debug("Found txid to bump: %s, ancestor_size = %d, ancestor_fee = %f", txid_to_bump, ancestor_size, ancestor_fee)
+
+        # Determine feerate of the last ancestor_size portion of the block.
+        cumulative_weight = 0
+        max_feerate = 0
+        for block_tx in reversed(template["transactions"]):
+            # The fee is the consensus correct one, not the policy modified
+            # one, but we haven't prioritized anything yet, so this should be fine.
+            if (4 * block_tx["fee"] / block_tx["weight"] > max_feerate):
+                max_feerate = 4 * block_tx["fee"] / block_tx["weight"]
+            cumulative_weight += block_tx["weight"]
+            if (cumulative_weight > 4*ancestor_size):
+                break
+
+        # Bump the candidate transaction so that it has a higher feerate than the
+        # max feerate observed in the last portion of the block that we seek to
+        # replace
+        fee_bump = int(max_feerate * ancestor_size - ancestor_fee) + 1
+
+        node.prioritisetransaction(txid_to_bump, fee_bump)
+
+        # Check that the next block template has txid_to_bump in it.
+        self.add_empty_block(node) # bypass the gbt cache
+        template = node.getblocktemplate({"rules":["segwit"]})
+        new_block_txids = [x["txid"] for x in template["transactions"]]
+
+        assert(txid_to_bump in new_block_txids)
+
+        # Get rid of our prioritisation
+        node.prioritisetransaction(txid_to_bump, -fee_bump)
+
+    def test_recent_transactions_excluded(self, node, recent_tx_window, recent_tx_stale_rate):
+        # Repeatedly try to call getblocktemplate, using mocktime to choose
+        # random points at which to call gbt
+        # If the block template contains recently-received transactions:
+        # - Prioritise all recent transactions to 0 fee, and then call gbt again
+        # - Verify that the fee decrease is more than the stale rate.
+        # Otherwise:
+        # - Verify that the block fee is within our tolerance of the highest fee
+        #   block we've seen so far.
+
+        # First, convert recent_tx_window from milliseconds to seconds.
+        recent_tx_window = int((recent_tx_window+999)/1000)
```

Comment:
In commit "Add functional test for CreateNewBlock"

Can use `//` operator for integer division in python